From 99091903bc8d5816a9b6b77fa2cb56a00bbd3270 Mon Sep 17 00:00:00 2001
From: Ran Wang <ran.wang_1@nxp.com>
Date: Thu, 9 Aug 2018 13:49:14 +0800
Subject: [PATCH 565/767] soc: fsl: Fix flex timer fail to wakeup on LS1021A

Correct the setting on ippdexpcr for flex timer alarm.
Use scratchpad register to pass setting to boot loader such as U-Boot,
because register ippdexpcr has defect of fail to return correct data
when read.

Signed-off-by: Ran Wang <ran.wang_1@nxp.com>
[Xulin: Original patch taken from NXP LSDK-18.09.]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/soc/fsl/layerscape/ftm_alarm.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/soc/fsl/layerscape/ftm_alarm.c b/drivers/soc/fsl/layerscape/ftm_alarm.c
index 4c79c02cc50a..ba590edb42e8 100644
--- a/drivers/soc/fsl/layerscape/ftm_alarm.c
+++ b/drivers/soc/fsl/layerscape/ftm_alarm.c
@@ -41,6 +41,7 @@
 
 static void __iomem *ftm1_base;
 static void __iomem *rcpm_ftm_addr;
+static void __iomem *scfg_scrachpad_addr;
 static u32 alarm_freq;
 static bool big_endian;
 
@@ -61,7 +62,7 @@ static struct rcpm_cfg ls1012a_rcpm_cfg = {
 
 static struct rcpm_cfg ls1021a_rcpm_cfg = {
 	.big_endian = BIG_ENDIAN,
-	.flextimer_set_bit = 0x20000,
+	.flextimer_set_bit = 0x30000000,
 };
 
 static struct rcpm_cfg ls1043a_rcpm_cfg = {
@@ -308,6 +309,12 @@ static int ftm_alarm_probe(struct platform_device *pdev)
 			iowrite32be(ippdexpcr, rcpm_ftm_addr);
 		else
 			iowrite32(ippdexpcr, rcpm_ftm_addr);
+
+		r = platform_get_resource_byname(pdev, IORESOURCE_MEM, "scrachpad");
+		if (r) {
+			scfg_scrachpad_addr = devm_ioremap_resource(&pdev->dev, r);
+			iowrite32(ippdexpcr, scfg_scrachpad_addr);
+		}
 	}
 
 	irq = irq_of_parse_and_map(np, 0);
-- 
2.17.0

